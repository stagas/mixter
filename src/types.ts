export type Class<T> = new(...args: any[]) => T

export type Ctor = new(...args: any[]) => unknown

export type Mixable<T extends Ctor> = {
  new(...args: any[]): InstanceType<T>
} & Omit<T, 'constructor'>

export type Mixin = Mixable<CustomElementConstructor>

/** @private */
export type CustomElementConstructor = Class<CustomElement> & {
  /**
   * The attribute keys in this list are observed by the DOM
   * and invoke {@link attributeChangedCallback} when they change.
   * @private
   */
  observedAttributes?: string[]
}

/** @private */
export interface CustomElement extends HTMLElement {
  /**
   * Callback that is invoked when one of the {@link withProperties} changes.
   *
   * @param name Name of attribute
   * @param oldValue Old value
   * @param newValue New value
   */
  attributeChangedCallback?(name: string, oldValue: string | null, newValue: string | null): void

  /**
   * Invoked when the component is added to the document's DOM.
   *
   * In `connectedCallback()` you should setup tasks that should only occur when
   * the element is connected to the document. The most common of these is
   * adding event listeners to nodes external to the element, like a keydown
   * event handler added to the window.
   *
   * ```ts
   * connectedCallback() {
   *   super.connectedCallback();
   *   this.addEventListener('keydown', this._handleKeydown);
   * }
   * ```
   *
   * Typically, anything done in `connectedCallback()` should be undone when the
   * element is disconnected, in `disconnectedCallback()`.
   */
  connectedCallback?(): void

  /**
   * Invoked when the component is removed from the document's DOM.
   *
   * This callback is the main signal to the element that it may no longer be
   * used. `disconnectedCallback()` should ensure that nothing is holding a
   * reference to the element (such as event listeners added to nodes external
   * to the element), so that it is free to be garbage collected.
   *
   * ```ts
   * disconnectedCallback() {
   *   super.disconnectedCallback();
   *   window.removeEventListener('keydown', this._handleKeydown);
   * }
   * ```
   *
   * An element may be re-connected after being disconnected.
   */
  disconnectedCallback?(): void
}

export type Root = {
  /** @private */
  root: ShadowRoot | HTMLElement
}
