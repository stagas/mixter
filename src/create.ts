import { Class, ContextFn, mixter, props, Root, shadow, state } from '.'

export const create = <P>(
  classOrFn: Class<P> | ContextFn<P & Root>,
  fn?: ContextFn<P & Root>,
) =>
  fn
    ? mixter(
      HTMLElement,
      shadow(),
      props(classOrFn as Class<P>),
      state(fn)
    )
    : mixter(HTMLElement, shadow(), state(classOrFn as ContextFn<Root>))
