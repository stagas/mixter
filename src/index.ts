export * from 'fluent-event'
export * from 'nested-css'
export * from './create'
export * from './mixins/attrs'
export * from './mixins/events'
export * from './mixins/lifecycle'
export * from './mixins/light'
export * from './mixins/props'
export * from './mixins/shadow'
export * from './mixins/state'
export * from './mixter'
export * from './types'
