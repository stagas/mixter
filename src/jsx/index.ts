export { fromElement as element } from 'html-vdom/from-element'
export type { Component } from 'html-vdom/from-element'
export * from './refs'
import type { Context, Fx } from '../mixins/state'
import { Root } from '../types'
import { hook, render } from './jsx-runtime'
import { renderer } from './renderer'

export const jsx = <T extends Root>($: Context<T>['$']) => ({
  ...renderer(render)($),
  part(fn: Fx<T, JSX.Element>['fn'], output?: any) {
    let update: any
    const Fn = () => {
      update = hook
      return output
    }
    $.effect(fn, value => {
      output = value
      update?.()
      return false
    })
    return Fn
  },
})
