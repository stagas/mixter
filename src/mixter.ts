import type { Ctor } from './types'

export function mixter<
  A extends Ctor,
  B extends Ctor,
  C extends Ctor,
  D extends Ctor,
  E extends Ctor,
  F extends Ctor,
  G extends Ctor,
  H extends Ctor,
  I extends Ctor,
  J extends Ctor,
  K extends Ctor,
>(
  a: A,
  b: (trait: A) => B,
  c: (trait: B) => C,
  d: (trait: C) => D,
  e: (trait: D) => E,
  f: (trait: E) => F,
  g: (trait: F) => G,
  h: (trait: G) => H,
  i: (trait: H) => I,
  j: (trait: I) => J,
  k: (trait: J) => K,
): A & B & C & D & E & F & G & H & I & J & K

export function mixter<
  A extends Ctor,
  B extends Ctor,
  C extends Ctor,
  D extends Ctor,
  E extends Ctor,
  F extends Ctor,
  G extends Ctor,
  H extends Ctor,
  I extends Ctor,
  J extends Ctor,
>(
  a: A,
  b: (trait: A) => B,
  c: (trait: B) => C,
  d: (trait: C) => D,
  e: (trait: D) => E,
  f: (trait: E) => F,
  g: (trait: F) => G,
  h: (trait: G) => H,
  i: (trait: H) => I,
  j: (trait: I) => J,
): A & B & C & D & E & F & G & H & I & J

export function mixter<
  A extends Ctor,
  B extends Ctor,
  C extends Ctor,
  D extends Ctor,
  E extends Ctor,
  F extends Ctor,
  G extends Ctor,
  H extends Ctor,
  I extends Ctor,
>(
  a: A,
  b: (trait: A) => B,
  c: (trait: B) => C,
  d: (trait: C) => D,
  e: (trait: D) => E,
  f: (trait: E) => F,
  g: (trait: F) => G,
  h: (trait: G) => H,
  i: (trait: H) => I,
): A & B & C & D & E & F & G & H & I

export function mixter<
  A extends Ctor,
  B extends Ctor,
  C extends Ctor,
  D extends Ctor,
  E extends Ctor,
  F extends Ctor,
  G extends Ctor,
  H extends Ctor,
>(
  a: A,
  b: (trait: A) => B,
  c: (trait: B) => C,
  d: (trait: C) => D,
  e: (trait: D) => E,
  f: (trait: E) => F,
  g: (trait: F) => G,
  h: (trait: G) => H,
): A & B & C & D & E & F & G & H

export function mixter<
  A extends Ctor,
  B extends Ctor,
  C extends Ctor,
  D extends Ctor,
  E extends Ctor,
  F extends Ctor,
  G extends Ctor,
>(
  a: A,
  b: (trait: A) => B,
  c: (trait: B) => C,
  d: (trait: C) => D,
  e: (trait: D) => E,
  f: (trait: E) => F,
  g: (trait: F) => G,
): A & B & C & D & E & F & G

export function mixter<
  A extends Ctor,
  B extends Ctor,
  C extends Ctor,
  D extends Ctor,
  E extends Ctor,
  F extends Ctor,
>(
  a: A,
  b: (trait: A) => B,
  c: (trait: B) => C,
  d: (trait: C) => D,
  e: (trait: D) => E,
  f: (trait: E) => F,
): A & B & C & D & E & F

export function mixter<A extends Ctor, B extends Ctor, C extends Ctor, D extends Ctor, E extends Ctor>(
  a: A,
  b: (trait: A) => B,
  c: (trait: B) => C,
  d: (trait: C) => D,
  e: (trait: D) => E,
): A & B & C & D & E

export function mixter<A extends Ctor, B extends Ctor, C extends Ctor, D extends Ctor>(
  a: A,
  b: (trait: A) => B,
  c: (trait: B) => C,
  d: (trait: C) => D,
): A & B & C & D

export function mixter<A extends Ctor, B extends Ctor, C extends Ctor>(
  a: A,
  b: (trait: A) => B,
  c: (trait: B) => C,
): A & B & C

export function mixter<A extends Ctor, B extends Ctor>(
  a: A,
  b: (trait: A) => B,
): A & B

export function mixter<A extends Ctor>(
  a: A,
  ...mixins: ((trait: any) => any)[]
) {
  return mixins.reduce((c, mixin) => mixin(c), a)
}
