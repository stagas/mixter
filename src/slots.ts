import { on } from 'fluent-event'

export const observeNodes = (
  nodes: Node[] | null,
  observer: MutationObserver,
  options: MutationObserverInit,
) => nodes?.forEach(node => observer.observe(node, options))

export const nodesToText = (nodes: Node[] | null) =>
  nodes
    ?.map(node => {
      const text = node.textContent!
      return text.trim().length ? text : ''
    })
    .join('') ?? ''

export type Slotted = { nodes: Node[]; elements: Element[]; firstChild?: Element }

export const slotted = (slots: HTMLSlotElement[]): Slotted => ({
  get nodes() {
    return slots.map(slot => slot.assignedNodes()).flat(Infinity) as Node[]
  },
  get elements() {
    return slots.map(slot => slot.assignedElements()).flat(Infinity) as Element[]
  },
  get firstChild() {
    return this.elements[0]
  },
})

export const onSlotChange = <T extends ShadowRoot | HTMLSlotElement>(
  el: T,
  cb: (slotted: Slotted) => void,
  fn: (el: T) => HTMLSlotElement[] = el => [...el.querySelectorAll('slot')],
) => on()(el, 'slotchange', () => cb(slotted(fn(el))))

export const onTextChange = <T extends ShadowRoot | HTMLSlotElement>(
  el: T,
  cb: (text: string) => void,
  fn?: (el: T) => HTMLSlotElement[],
) => {
  let observer: MutationObserver

  const off = onSlotChange(el, ({ nodes }) => {
    observer?.disconnect()
    if (nodes.length) {
      observer = new MutationObserver(() => cb(nodesToText(nodes)))
      observeNodes(nodes, observer, { characterData: true })
      cb(nodesToText(nodes))
    }
  }, fn)

  return () => {
    observer?.disconnect()
    off()
  }
}
