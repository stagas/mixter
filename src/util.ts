export const accessors = (target: any, source: any, fn: (key: any, value: any) => PropertyDescriptor) =>
  Object.defineProperties(
    target,
    Object.fromEntries(
      Object.entries(source)
        .map(
          ([key, value]) => {
            const next = fn(key, value) as any
            const prev = Object.getOwnPropertyDescriptor(target, key)
            if (prev && prev.get && prev.set) {
              const { get, set } = next
              next.get = () => (prev.get?.() ?? get())
              next.set = (v: any) => {
                prev.set?.(v)
                set(prev.get?.() ?? v)
              }
            }
            return [key, {
              configurable: true,
              enumerable: true,
              ...next,
            }]
          }
        )
    )
  )

export const kebab = (s: string) => s.replace(/[a-z](?=[A-Z])|[A-Z]+(?=[A-Z][a-z])/g, '$&-').toLowerCase()
