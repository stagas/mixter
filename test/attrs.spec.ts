import { attrs, mixter } from '../src'

const create = (Ctor: CustomElementConstructor) => {
  const randTag = 'x-' + ((Math.random() * 10e6) | 0).toString(36)
  customElements.define(randTag, Ctor)
  return randTag
}

describe('withProps', () => {
  it('creates a base class that mixins a given props instance', () => {
    class Foo extends mixter(
      HTMLElement,
      attrs(
        class {
          foo = String
          bar = String
        }
      )
    ) {}

    create(Foo)

    const el = new Foo()

    expect(el.foo).toBeUndefined()
    el.setAttribute('foo', 'some value')
    expect(el.foo).toEqual('some value')

    expect(el.bar).toBeUndefined()
    el.setAttribute('bar', 'some other value')
    expect(el.bar).toEqual('some other value')
  })

  it('fires propertyChangedCallback', () => {
    const propChangedResults: any[] = []

    class Foo extends mixter(
      HTMLElement,
      attrs(
        class {
          fooFoo = String
          barBar = String
        }
      )
    ) {
      propertyChangedCallback(name: string, oldValue: any, newValue: any) {
        propChangedResults.push({ name, oldValue, newValue })
      }
    }

    create(Foo)

    const el = new Foo()

    expect(el.fooFoo).toBeUndefined()
    el.setAttribute('foofoo', 'some value')
    expect(el.fooFoo).toEqual('some value')
    el.fooFoo = 'another'
    expect(el.fooFoo).toEqual('another')
    el.fooFoo = 'another'

    expect(el.barBar).toBeUndefined()
    el.setAttribute('barbar', 'some other value')
    expect(el.barBar).toEqual('some other value')

    expect(propChangedResults).toMatchSnapshot()
  })

  it('keys are enumerable', () => {
    class Foo extends mixter(
      HTMLElement,
      attrs(
        class {
          foo = String
          bar = String
        }
      )
    ) {}

    create(Foo)

    const el = new Foo()

    expect(Object.keys(el)).toEqual(['foo', 'bar'])
  })

  it('converts kebab-case attribute names to propCase camelCased ones', () => {
    class Foo extends mixter(
      HTMLElement,
      attrs(
        class {
          fooFoo = String
          barBar = String
        }
      )
    ) {}

    create(Foo)

    const el = new Foo()

    expect(el.fooFoo).toBeUndefined()
    el.setAttribute('foofoo', 'some value')
    expect(el.fooFoo).toEqual('some value')

    expect(el.barBar).toBeUndefined()
    el.setAttribute('barbar', 'some other value')
    expect(el.barBar).toEqual('some other value')
  })

  it('works with inheritance', () => {
    class Foo extends mixter(
      HTMLElement,
      attrs(
        class {
          foo = String
        }
      )
    ) {}

    class Bar extends mixter(
      Foo,
      attrs(
        class {
          bar = String
        }
      )
    ) {}

    create(Bar)

    const el = new Bar()

    expect(el.foo).toBeUndefined()
    el.setAttribute('foo', 'some value')
    expect(el.foo).toEqual('some value')

    expect(el.bar).toBeUndefined()
    el.setAttribute('bar', 'some other value')
    expect(el.bar).toEqual('some other value')
  })

  it('casts values to primitive types', () => {
    class Foo extends mixter(
      HTMLElement,
      attrs(
        class {
          string? = String
          number? = Number
          boolean = Boolean
          implicitString = 'string'
          implicitNumber = 123
          implicitBoolean = true
          // somethingElse? = new Uint8Array(1)
        }
      )
    ) {}

    create(Foo)

    const el = new Foo()

    expect(el.string).toBeUndefined()
    expect(el.number).toBeUndefined()
    expect(el.boolean).toBeUndefined()

    el.string = 123 as any
    expect(el.string).toBe('123')
    el.setAttribute('string', '456')
    expect(el.string).toBe('456')

    el.number = '123' as any
    expect(el.number).toBe(123)
    el.setAttribute('number', '456')
    expect(el.number).toBe(456)

    el.boolean = '' as any
    expect(el.boolean).toBe(true)
    el.setAttribute('boolean', '')
    expect(el.boolean).toBe(true)
    el.toggleAttribute('boolean')
    expect(el.boolean).toBe(false)

    expect(el.implicitString).toBe('string')
    expect(el.implicitNumber).toBe(123)
    expect(el.implicitBoolean).toBe(true)

    el.implicitString = 123 as any
    expect(el.implicitString).toBe('123')

    el.implicitNumber = '123' as any
    expect(el.implicitNumber).toBe(123)
    el.setAttribute('implicitnumber', '456')
    expect(el.implicitNumber).toBe(456)

    el.implicitBoolean = '' as any
    expect(el.implicitBoolean).toBe(true)
    el.setAttribute('implicitboolean', '')
    expect(el.implicitBoolean).toBe(true)
    el.toggleAttribute('implicitboolean')
    expect(el.implicitBoolean).toBe(false)

    // el.setAttribute('something-else', '123')
    // expect(el.somethingElse).toBeInstanceOf(Uint8Array)
  })

  // it('uses `schema` static props class property when detected', () => {
  //   class Foo extends withProperties(
  //     HTMLElement,
  //     class {
  //       static schema = {
  //         foo: '',
  //         bar: '',
  //       }
  //       foo = 'some default'
  //     }
  //   ) {}

  //   create(Foo)

  //   const el = new Foo()

  //   expect(el.foo).toEqual('some default')
  //   el.setAttribute('foo', 'some value')
  //   expect(el.foo).toEqual('some value')

  //   expect(el.bar).toBeUndefined()
  //   el.setAttribute('bar', 'some other value')
  //   expect(el.bar).toEqual('some other value')
  // })
})
