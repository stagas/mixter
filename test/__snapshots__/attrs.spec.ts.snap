// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withProps fires propertyChangedCallback 1`] = `
Array [
  Object {
    "name": "fooFoo",
    "newValue": "some value",
    "oldValue": undefined,
  },
  Object {
    "name": "fooFoo",
    "newValue": "another",
    "oldValue": "some value",
  },
  Object {
    "name": "barBar",
    "newValue": "some other value",
    "oldValue": undefined,
  },
]
`;
